mrp_lot_number_propagation/models/mrp_bom_line.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2022 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class MrpBomLine(models.Model):
+    _inherit = "mrp.bom.line"
+
+    propagate_lot_number = fields.Boolean(
+        default=False,
+    )
+    display_propagate_lot_number = fields.Boolean(
+        compute="_compute_display_propagate_lot_number"
+    )
+
+    @api.depends(
+        "bom_id.display_lot_number_propagation",
+        "bom_id.lot_number_propagation",
+    )
+    def _compute_display_propagate_lot_number(self):
+        for line in self:
+            line.display_propagate_lot_number = (
+                line.bom_id.display_lot_number_propagation
+                and line.bom_id.lot_number_propagation
+            )
+
+    @api.constrains(
+        "propagate_lot_number",
+        "bom_id.lot_number_propagation",
+        "product_id.tracking",
+        "bom_id.product_tmpl_id.tracking",
```

Comment:
I added some constraints on `product.product`/`product.template`/`mrp.bom` to avoid this situation.
If we tick "Lot Number Propagation" on a BoM, it's mandatory to tick "Propagate Lot Number" on a BoM line to not let the BoM in a wobbly state. Also, if a product is related to a BoM or BoM line, its tracking can't be unset (the user will have to update/delete/archive related BoM before).